core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -71,12 +75,22 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, scheduler: TaskSchedule
   }
   
   override def receiveWithLogging: PartialFunction[Any, Unit] = {
-    case Heartbeat(executorId, taskMetrics, blockManagerId) =>
-      val unknownExecutor = !scheduler.executorHeartbeatReceived(
-        executorId, taskMetrics, blockManagerId)
-      val response = HeartbeatResponse(reregisterBlockManager = unknownExecutor)
-      executorLastSeen(executorId) = System.currentTimeMillis()
-      sender ! response
+    case RegisterTaskScheduler(scheduler) =>
+      this.scheduler = scheduler
+    case heartbeat @ Heartbeat(executorId, taskMetrics, blockManagerId) =>
+      if (scheduler != null) {
+        val unknownExecutor = !scheduler.executorHeartbeatReceived(
+          executorId, taskMetrics, blockManagerId)
+        val response = HeartbeatResponse(reregisterBlockManager = unknownExecutor)
+        executorLastSeen(executorId) = System.currentTimeMillis()
+        sender ! response
+      } else {
+        // Because Executor will sleep several seconds then send the first "Heartbeat", this case
+        // rarely happens. However, if it really happens, log it and ask the executor to register
+        // itself again.
+        logWarning(s"Dropping $heartbeat because TaskScheduler has not been ready yet")
+        sender ! HeartbeatResponse(reregisterBlockManager = true)
```

Comment:
For `reregistering`: Because Executor uses `askWithReply` to send `Heartbeat`, I want to send something back to Executor to avoid timeout.